src/pages/home/ReportScreen.js

```diff
@@ -176,7 +199,15 @@ function ReportScreen({
         Performance.markStart(CONST.TIMING.CHAT_RENDER);
     }
 
-    const reportID = getReportID(route);
+    // Define here because reportActions are recalculated before mount, allowing data to display faster than useEffect can trigger. If we have cached reportActions, they will be shown immediately. We aim to display a loader first, then fetch relevant reportActions, and finally show them.
+    useMemo(() => {
```

Comment:
Next up, `isLinkingToMessage`. If I'm correct in my understanding, `isLinkingToMessage` is `true` if:

- there's a `reportActionID` in the URL, AND
- `reportMetadata.isLoadingInitialReportActions` is `true`

Otherwise it's `false`. So I don't think this needs to be a state variable! It can just be replaced with `!!reportActionID && reportMetadata.isLoadingInitialReportActions`.

It's only read in `shouldShowSkeleton`, so if we replace that inline:

```diff
- isLinkingToMessage || !isReportReadyForDisplay || reportMetadata.isLoadingInitialReportActions || isLoading
+ (!!reportActionID && reportMetadata.isLoadingInitialReportActions) || !isReportReadyForDisplay || reportMetadata.isLoadingInitialReportActions || isLoading
```

That in turn [can be simplified](https://www.wolframalpha.com/input?i=%28A+AND+B%29+OR+C+OR+B+OR+D) like so:

```js
const shouldShowSkeleton = useMemo(
        () => reportMetadata.isLoadingInitialReportActions || !isReportReadyForDisplay || isLoading,
        [reportMetadata.isLoadingInitialReportActions, isReportReadyForDisplay, isLoading],
    );
```

Then we can remove `isLinkingToMessage` ✅

Reply:
Next up, `shouldTriggerLoadingRef`. This value was only ever read as a way to control either itself or `isLinkingToMessage`. Since `isLinkingToMessage` has been removed, now we can remove `shouldTriggerLoadingRef` as well. ✅

Reply:
At this point, we have only the following remaining:

- `reportMetadata.isLoadingInitialReportActions`
- `isSidebarLoaded`
- `isLoading`
- `isReportReadyForDisplay`

All of which seem slightly different. Depending on if the `reportID` from Onyx can indeed be different from the `reportID` from route params, it might be possible to simplify further. But already, this is a huge improvement. Here's what I have:

```diff
diff --git a/src/pages/home/ReportScreen.js b/src/pages/home/ReportScreen.js
index 312a02dfa1..12d009096e 100644
--- a/src/pages/home/ReportScreen.js
+++ b/src/pages/home/ReportScreen.js
@@ -173,10 +173,8 @@ function ReportScreen({
     const flatListRef = useRef();
     const reactionListRef = useRef();
     const firstRenderRef = useRef(true);
-    const shouldTriggerLoadingRef = useRef(!!reportActionID);
     const prevReport = usePrevious(report);
     const prevUserLeavingStatus = usePrevious(userLeavingStatus);
-    const [isLinkingToMessage, setLinkingToMessage] = useState(!!reportActionID);
 
     const reportActions = useMemo(() => {
         if (!!allReportActions && allReportActions.length === 0) {
@@ -199,21 +197,9 @@ function ReportScreen({
         Performance.markStart(CONST.TIMING.CHAT_RENDER);
     }
 
-    // Define here because reportActions are recalculated before mount, allowing data to display faster than useEffect can trigger. If we have cached reportActions, they will be shown immediately. We aim to display a loader first, then fetch relevant reportActions, and finally show them.
-    useMemo(() => {
-        shouldTriggerLoadingRef.current = !!reportActionID;
-        // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [route, reportActionID]);
-    useLayoutEffect(() => {
-        setLinkingToMessage(!!reportActionID);
-    }, [route, reportActionID]);
-
     const {addWorkspaceRoomOrChatPendingAction, addWorkspaceRoomOrChatErrors} = ReportUtils.getReportOfflinePendingActionAndErrors(report);
     const screenWrapperStyle = [styles.appContent, styles.flex1, {marginTop: viewportOffsetTop}];
 
-    // There are no reportActions at all to display and we are still in the process of loading the next set of actions.
-    const isLoadingInitialReportActions = _.isEmpty(reportActions) && reportMetadata.isLoadingInitialReportActions;
-
     const isOptimisticDelete = lodashGet(report, 'statusNum') === CONST.REPORT.STATUS.CLOSED;
 
     const shouldHideReport = !ReportUtils.canAccessReport(report, policies, betas);
@@ -309,12 +295,12 @@ function ReportScreen({
         // It possible that we may not have the report object yet in Onyx yet e.g. we navigated to a URL for an accessible report that
         // is not stored locally yet. If report.reportID exists, then the report has been stored locally and nothing more needs to be done.
         // If it doesn't exist, then we fetch the report from the API.
-        if (report.reportID && report.reportID === getReportID(route) && !isLoadingInitialReportActions) {
+        if (report.reportID && report.reportID === getReportID(route) && !reportMetadata.isLoadingInitialReportActions) {
             return;
         }
 
         fetchReport();
-    }, [report.reportID, route, isLoadingInitialReportActions, fetchReport]);
+    }, [report.reportID, route, reportMetadata.isLoadingInitialReportActions, fetchReport]);
 
     const dismissBanner = useCallback(() => {
         setIsBannerVisible(false);
@@ -504,22 +490,10 @@ function ReportScreen({
     const actionListValue = useMemo(() => ({flatListRef, scrollPosition, setScrollPosition}), [flatListRef, scrollPosition, setScrollPosition]);
 
     const shouldShowSkeleton = useMemo(
-        () => isLinkingToMessage || !isReportReadyForDisplay || isLoadingInitialReportActions || isLoading || (!!reportActionID && reportMetadata.isLoadingInitialReportActions),
-        [isLinkingToMessage, isReportReadyForDisplay, isLoadingInitialReportActions, isLoading, reportActionID, reportMetadata.isLoadingInitialReportActions],
+        () => reportMetadata.isLoadingInitialReportActions || !isReportReadyForDisplay || isLoading,
+        [reportMetadata.isLoadingInitialReportActions, isReportReadyForDisplay, isLoading],
     );
 
-    // This helps in tracking from the moment 'route' triggers useMemo until isLoadingInitialReportActions becomes true. It prevents blinking when loading reportActions from cache.
-    useEffect(() => {
-        if (reportMetadata.isLoadingInitialReportActions && shouldTriggerLoadingRef.current) {
-            shouldTriggerLoadingRef.current = false;
-            return;
-        }
-        if (!reportMetadata.isLoadingInitialReportActions && !shouldTriggerLoadingRef.current) {
-            shouldTriggerLoadingRef.current = false;
-            setLinkingToMessage(false);
-        }
-    }, [reportMetadata.isLoadingInitialReportActions]);
-
     return (
         <ActionListContext.Provider value={actionListValue}>
             <ReactionListContext.Provider value={reactionListRef}>

```

Reply:
Also, I suggest renaming `reportMetadata.isLoadingInitialReportActions` to `isLoadingReportActions`. Initially, `isLoadingInitialReportActions` indicated that it was the first render of the report, meaning there were no reproductions for this report. Now we call it when back online and in a bunch of other places.
What do you think?

Reply:
> I suggest renaming reportMetadata.isLoadingInitialReportActions to isLoadingReportActions

Let's not for now. I think even during reconnect `isLoadingInitialReportActions` makes sense, especially when juxtaposed against `isLoadingOlderReportActions` and `isLoadingNewerReportActions`. Also, this PR is big enough as-is.